net.sf.eclipsecs.ui/src/net/sf/eclipsecs/ui/quickfixes/MarkerHelpContextProvider.java

```diff
@@ -0,0 +1,161 @@
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+//
+//============================================================================
+
+package net.sf.eclipsecs.ui.quickfixes;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.ui.IMarkerHelpContextProvider;
+
+import net.sf.eclipsecs.core.builder.CheckstyleMarker;
+import net.sf.eclipsecs.core.config.meta.MetadataFactory;
+
+import org.eclipse.help.AbstractContextProvider;
+import org.eclipse.help.IContext;
+import org.eclipse.help.IHelpResource;
+
+/**
+ * Provides context help for checkstyle markers.
+ */
+public class MarkerHelpContextProvider extends AbstractContextProvider
+        implements IMarkerHelpContextProvider {
+
+  /**
+   * package of the regexp checks
+   */
+  private static final String REGEXP_PACKAGE = "com.puppycrawl.tools.checkstyle.checks.regexp.";
+  /**
+   * suffix of all standard check implementations
+   */
+  private static final String CHECK_SUFFIX = "Check";
+  /**
+   * Common prefix for all Checkstyle marker help contexts. Must be same as plugin id and must end
+   * with a dot.
+   */
+  private static final String PLUGIN_PREFIX = "net.sf.eclipsecs.ui" + ".";
+
+  @Override
+  public String getHelpContextForMarker(IMarker marker) {
+    String module = getModule(marker);
+    if (!module.endsWith(CHECK_SUFFIX)) {
+      return null;
+    }
+    return PLUGIN_PREFIX + StringUtils.removeEnd(StringUtils.substringAfterLast(module, '.'), CHECK_SUFFIX);
+  }
+
+  private String getModule(IMarker marker) {
+    return marker.getAttribute(CheckstyleMarker.MODULE_NAME, StringUtils.EMPTY);
+  }
+
+  @Override
+  public boolean hasHelpContextForMarker(IMarker marker) {
+    if (!CheckstyleMarker.isCheckstyleMarker(marker)) {
+      return false;
+    }
+    if (getModule(marker).startsWith(REGEXP_PACKAGE)) {
+      // regex rules don't provide useful help for understanding and fixing an issue
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public IContext getContext(String id, String locale) {
+    var moduleName = StringUtils.substringAfter(id, PLUGIN_PREFIX);
+    return new CheckstyleHelpContext(moduleName);
+  }
+
+  @Override
+  public String[] getPlugins() {
+    return new String[] { "net.sf.eclipsecs.ui" };
+  }
+
+  /**
+   * @param moduleName
+   *          module name
+   * @return online help URL
+   */
+  public static String getOnlineHelp(String moduleName) {
+    var metadata = MetadataFactory.getRuleMetadata(moduleName);
+    if (metadata == null) {
+      return null;
+    }
+    var group = metadata.getGroup().getGroupId().toLowerCase();
+    // some web pages are different to the packages in Checkstyle
+    if ("indentation".equals(group)) {
+      group = "misc";
+    }
+    var file = moduleName.toLowerCase();
+    return "https://checkstyle.org/checks/" + group + "/" + file + ".html#" + moduleName;
```

Comment:
I am not sure all pages have the `#` and would ensure that is the case.

I know in the beginning there was discussions on this and I was against it since we only have 1 module per page now, so the `#` seemed excessively unnecessary and was really only skipping our page header. If we don't like our header that much, we might as well remove it.

Reply:
I am going to resolve this. I believe the actual issue was the links on our group pages. All the pages have the `#` because of how we define the page and the `#` creation.

Reply:
Yeah, I think I know your confusion. I had similar when looking for the scheme. If you navigate from the TOC of the checkstyle page, the very first link of each group is different than the remaining ones. However, even for the first one of each group, the same link _exists_, so it can be used here.